table/pit-cs-tree.go

```diff
@@ -0,0 +1,382 @@
+		}
+	}
+	return matching
+}
+
+func (p *PitCsTree) FindInterestPrefixMatchByData(data *ndn.Data, token *uint32) []PitEntry {
+	if token != nil {
+		if entry, ok := p.pitTokenMap[*token]; ok && entry.Token() == *token {
+			return []PitEntry{entry}
+		}
+		return nil
+	}
+
+	matching := make([]PitEntry, 0)
+	dataNameLen := data.Name().Size()
+	for curNode := p.root.findLongestPrefixEntry(data.Name()); curNode != nil; curNode = curNode.parent {
+		for _, entry := range curNode.pitEntries {
+			if entry.CanBePrefix() || curNode.depth == dataNameLen {
+				matching = append(matching, entry)
+			}
+		}
+	}
+	return matching
+}
+
+// PitSize returns the number of entries in the PIT.
+func (p *PitCsTree) PitSize() int {
+	return p.nPitEntries
+}
+
+// CsSize returns the number of entries in the CS.
+func (p *PitCsTree) CsSize() int {
+	return p.nCsEntries
+}
+
+// IsCsAdmitting returns whether the CS is admitting contents.
+func (p *PitCsTree) IsCsAdmitting() bool {
+	return csAdmit
+}
+
+// IsCsServing returns whether the CS is serving contents.
+func (p *PitCsTree) IsCsServing() bool {
+	return csServe
+}
+
+// Originally implemented as FindOrInsertOutRecord()
+func (e *nameTreePitEntry) InsertOutRecord(interest *ndn.Interest, face uint64) *PitOutRecord {
+	var record *PitOutRecord
+	var ok bool
+	if record, ok = e.outRecords[face]; !ok {
+		record := new(PitOutRecord)
+		record.Face = face
+		record.LatestNonce = interest.Nonce()
+		record.LatestTimestamp = time.Now()
+		record.LatestInterest = interest
+		record.ExpirationTime = time.Now().Add(interest.Lifetime())
+		e.outRecords[face] = record
+		return record
+	}
+
+	// Existing record
+	record.LatestNonce = interest.Nonce()
+	record.LatestTimestamp = time.Now()
+	record.LatestInterest = interest
+	record.ExpirationTime = time.Now().Add(interest.Lifetime())
+	return record
+}
+
+// Originally implemented as FindOrInsertOutRecord()
+func (e *nameTreePitEntry) GetOutRecords() []*PitOutRecord {
+	records := make([]*PitOutRecord, 0)
+	for _, value := range e.outRecords {
+		records = append(records, value)
+	}
+	return records
+}
+
+func (p *pitCsTreeNode) findExactMatchEntry(name *ndn.Name) *pitCsTreeNode {
+	if name.Size() > p.depth {
+		for _, child := range p.children {
+			if name.At(child.depth - 1).Equals(child.component) {
+				return child.findExactMatchEntry(name)
+			}
+		}
+	} else if name.Size() == p.depth {
+		return p
+	}
+	return nil
+}
+
+func (p *pitCsTreeNode) findLongestPrefixEntry(name *ndn.Name) *pitCsTreeNode {
+	if name.Size() > p.depth {
+		for _, child := range p.children {
+			if name.At(child.depth - 1).Equals(child.component) {
+				return child.findLongestPrefixEntry(name)
+			}
+		}
+	}
+	return p
+}
+
+func (p *pitCsTreeNode) fillTreeToPrefix(name *ndn.Name) *pitCsTreeNode {
+	curNode := p.findLongestPrefixEntry(name)
+	for depth := curNode.depth + 1; depth <= name.Size(); depth++ {
+		newNode := new(pitCsTreeNode)
+		newNode.component = name.At(depth - 1).DeepCopy()
```

Comment:
FYI: The original code abuses DeepCopy and make tons of unnecessary copies. I don't know if they are really necessary. But for now probably leaves it there is a good choice, because we don't want to break functionality unintendedly during refactoring.